state/fork/traverse.go

```diff
@@ -0,0 +1,117 @@
+package fork
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+)
+
+// functor that will be called on each block header when traversing blocks.
+type onVisitBlock = func(header *flow.Header) error
+
+// TraverseBackward traverses the given fork (specified by block ID `forkHead`)
+// in the order of decreasing height. The `terminal` defines when the traversal
+// stops. The `visitor` callback is called for each block in this segment.
+func TraverseBackward(headers storage.Headers, forkHead flow.Identifier, visitor onVisitBlock, terminal Terminal) error {
+	lowestHeightToVisit, sanityChecker, err := terminal.Translate2Height(headers)
+	if err != nil {
+		return fmt.Errorf("error evaluating terminal: %w", err)
+	}
+
+	startBlock, err := headers.ByBlockID(forkHead)
+	if err != nil {
+		return fmt.Errorf("could not fork head with if  %x: %w", forkHead, err)
+	}
+	if startBlock.Height < lowestHeightToVisit {
+		return nil
+	}
+
+	return unsafeTraverse(headers, startBlock, visitor, lowestHeightToVisit, sanityChecker)
+}
+
+// unsafeTraverse implements the fork traversal in the order of decreasing height.
+// It is unsafe because:
+// * always calls the `visitor` on the `block`, _before_ it checks whether to stop
+// * if `block` has a lower height than lowestHeightToVisit, the traversal keeps
+//   going until we hit the root block
+func unsafeTraverse(headers storage.Headers, block *flow.Header, visitor onVisitBlock, lowestHeightToVisit uint64, sanityChecker sanityCheckLowestVisitedBlock) error {
+	for {
+		err := visitor(block)
+		if err != nil {
+			return err
+		}
+
+		if block.Height == lowestHeightToVisit {
+			err = sanityChecker(block)
+			if err != nil {
+				return fmt.Errorf("lowest visited block at height %d lowestHeightToVisit failed sanity check: %w", lowestHeightToVisit, err)
+			}
+			return nil
+		}
+
+		block, err = headers.ByBlockID(block.ParentID)
+		if err != nil {
+			return fmt.Errorf("could not get block header %x: %w", block.ParentID, err)
+		}
+	}
+}
+
+// TraverseForward traverses the given fork (specified by block ID `forkHead`)
+// in the order of increasing height. The `terminal` defines when the traversal
+// stops. The `visitor` callback is called for each block in this segment.
```

Comment:
```suggestion
// TraverseForward traverses the given fork (specified by block ID `forkHead`)
// in the order of increasing height. The `terminal` defines when the traversal
// begins. The `visitor` callback is called for each block in this segment.
```
If we are saying the traversal is in increasing height order, the terminal is defining the beginning of the traversal. 

The more I think about the `Forward` version of this API, the more I wonder if it would be simpler to just provide an API to get a range of blocks, then let the client decide how to iterate it. The main benefit of using the callback for the `TraverseBackward` case is that we can fail fast and skip the rest of the traversal if an error occurs or we hit a dynamic halting condition, but if we're traversing forward we're paying for the whole traversal anyway.
```go
func HeightRange(from, to uint64) ([]*flow.Header, error) { ... }

func Range(from, to flow.Identifier) ([]*flow.Header, error) { ... }
```

Reply:
Thanks for the documentation revision. 

Regarding the `Range` suggestion, I agree that this is nice for traversing forward. Though, in the dominant number of cases, we are actually traversing backwards. To me, this seems like we have three options:
1. use `Range` everywhere, which means we are paying the cost of unnecessarily allocating the slice `[]*flow.Header`  for traversing backwards (most of the cases)
2. use different APIs depending on direction `TraverseBackward` and `Range` 
3. use consistent API (i.e. `TraverseBackward` and `TraverseForward`) 

While I opted for 3., I totally agree that it is a matter of personal preference. Nevertheless, I think that 3. is the most complex to implement, but offers the combination of performance and API consistency.